exonum-java-binding-core/src/main/java/com/exonum/binding/messages/TransactionExecutionException.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2018 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exonum.binding.messages;
+
+import javax.annotation.Nullable;
+
+/**
+ * An exception occurred during transaction execution. The transaction exception includes
+ * an integer error code, that may be either service-specific or transaction-specific;
+ * and an optional description — an exception message. Both the error code and the description
+ * are saved in the database, but only the value of the error code affects the blockchain state.
+ *
+ * <p>The other attributes of a Java exception — a stack trace, a cause, suppressed exceptions —
+ * are not saved in the database and are used for logging only.
+ *
+ * <p>An external client will get the error code and description when requests a transaction
+ * status of a failed transaction. See
+ * <a href="https://exonum.com/doc/advanced/node-management/#transaction">the API endpoint documentation</a>
+ * for more information.
+ */
+public class TransactionExecutionException extends Exception {
+
+  // TODO: Consider using enums and taking their ordinal as the error code?
```

Comment:
WDYT? Won't be able to assign arbitrary values to enum elements, or skip some values, but it will be somewhat easier to work with.

Reply:
In core user defines his own enums and converts them into error codes.
https://github.com/exonum/exonum/blob/7ba9b1ed739c7baf0a13c6f70a8fca81ce6c5930/examples/cryptocurrency/src/lib.rs#L165-L199
I don't understand what you suggest to do in Java.

Reply:
Passing any Enum and taking its position as the error code: https://docs.oracle.com/javase/10/docs/api/java/lang/Enum.html#ordinal()

We can require an Enum implementing a certain interface (like this with `short id()`: https://github.com/exonum/exonum-java-binding/blob/83010162593fb44f26677a7ed13fd4dcf36b5783/exonum-java-binding-qa-service/src/main/java/com/exonum/binding/qaservice/transactions/QaTransaction.java#L27), but it would be a bigger constraint.

Reply:
Ah, I see know, thanks. Actually, idea of enforcing enums usage seems good to me.

Reply:
I am not sure we shall enforce that because if you have a single or a couple of error codes, defining an enum might seem too verbose (especially if that's an enum implementing a certain interface with `byte errorCode()` method). If we just use ordinals then it's OK in terms of verbosity, but the error codes are (a) implicit; (b) can no longer be arbitrary.

Reply:
@bullet-tooth , @MakarovS , wdyt?

Reply:
https://jira.bf.local/browse/ECR-2006

Reply:
I also like to use enum here. As example we can handle user defined range of errors by some enum instance:
```
public enum ErrorEnum {
  SYSTEM_ERROR,
  CUSTOM_ERROR,
  DEFAULT_ERROR;
  
  public static ErrorEnum fromExonumErrorCode(int errorCode) {
    if(errorCode == -1) {
      return SYSTEM_ERROR;
    } else if(errorCode > 100 && errorCode <= 200) {
      return CUSTOM_ERROR;
    }
    return DEFAULT_ERROR;
  }
  
}
```

Reply:
yes, but as far as I know there is a codes range for such errors. And my point was this range can be mapped to one enum value

Reply:
I see, but how do you pass that value (= errorCode as a byte) to the native code, which does not understand enums? 

e.g.
```java
if (foo) {
  throw new ExecutionException(FOO);
} else if (bar) {
  throw new ExecutionException(BAR);
}
```

FOO & BAR may be:
* byte constants;
* simple enum constants (you translate them to a byte error code using their ordinal `int ordinal()`);
* enum constants implementing a certain interface (`byte toErrorCode()`).